prover/zkevm/prover/common/flatten_column.go

```diff
@@ -0,0 +1,203 @@
+		limbs:             flattenLimbs,
+		mask:              comp.InsertCommit(0, ifaces.ColIDf("%s.%s_FLATTEN_MASK", module, circuit), flattenSize),
+		auxProjectionMask: auxProjectionMask,
+		nbLimbsCols:       nbLimbsCols,
+		onesColumn:        onesColumn,
+		module:            module,
+		circuit:           circuit,
+		isDuplicated:      isDuplicated,
+	}
+}
+
+// Limbs returns the flattened limbs column.
+func (l *FlattenColumn) Limbs() ifaces.Column {
+	return l.limbs
+}
+
+// Mask returns the flattened mask column.
+func (l *FlattenColumn) Mask() ifaces.Column {
+	return l.mask
+}
+
+// CsFlattenProjection adds a single batched projection constraint that enforces
+// the “flattened” limbs and mask columns exactly match the row‐wise concatenation
+// of the original limb columns and their mask.
+//
+// It works by:
+//  1. Shifting the committed flattened limbs and mask by 0…nbLimbsCols−1 rows.
+//  2. Duplicating the original limbs and mask into nbLimbsCols slots.
+//  3. Using auxProjectionMask (a sparse selector with a 1 at each block start)
+//     and a onesColumn to feed into the projection gadget.
+//  4. Requiring at each selected position that
+//     shiftedFlattenLimbs[i] == originalLimbs[i]
+//     shiftedFlattenMask[i]  == originalMask
+//     for all limb indices i and row positions.
+//
+// Suppose nbLimbsCols = 3, size = 4:
+// original limbs (rows × limbs):
+//
+//	r\i   0    1    2
+//	 0  [a0,  b0,  c0]
+//	 1  [a1,  b1,  c1]
+//	 2  [a2,  b2,  c2]
+//	 3  [a3,  b3,  c3]
+//
+// flattened (size*3 = 12):
+//
+//	[a0, b0, c0,  a1, b1, c1,  a2, b2, c2,  a3, b3, c3]
+//
+// shift0 (i=0):
+//
+//	[a0, b0, c0,  a1, b1, c1,  a2, b2, c2,  a3, b3, c3]
+//
+// shift1 (i=1):
+//
+//	[b0, c0, a1,  b1, c1, a2,  b2, c2, a3,  b3, c3,  0 ]
+//
+// shift2 (i=2):
+//
+//	[c0, a1, b1,  c1, a2, b2,  c2, a3, b3,  c3,  0,   0 ]
+//
+// auxProjectionMask   = [1,0,0, 1,0,0, 1,0,0, 1,0,0]
+//
+// At each ‘1’ at idx = r*3, for shift i, enforce:
+//
+//	shift_i[idx] == original[r][i]
+//
+// CsFlattenProjection batches all these equalities into one projection check.
+func (l *FlattenColumn) CsFlattenProjection(comp *wizard.CompiledIOP, limbs []ifaces.Column, mask ifaces.Column) {
+	masks := make([]ifaces.Column, l.nbLimbsCols)
+	shiftedFlattenLimbs := make([]ifaces.Column, l.nbLimbsCols)
+	shiftedFlattenMask := make([]ifaces.Column, l.nbLimbsCols)
+
+	for i := 0; i < l.nbLimbsCols; i++ {
+		masks[i] = mask
+		shiftedFlattenLimbs[i] = column.Shift(l.limbs, i)
+		shiftedFlattenMask[i] = column.Shift(l.mask, i)
+	}
+
+	projection.InsertProjection(comp, ifaces.QueryIDf("%v_%s_FLATTEN_PROJECTION", l.module, l.circuit),
+		append(shiftedFlattenLimbs[:], shiftedFlattenMask[:]...),
+		append(limbs[:], masks[:]...),
+		l.auxProjectionMask, l.onesColumn,
+	)
+
+	l.originalMask = mask
+	l.originalLimbs = limbs
+}
+
+// Assign maps trace limb columns and mask into the flattened columns.
+func (l *FlattenColumn) Assign(run *wizard.ProverRuntime) {
```

Comment:
Minor note; if you rename the function [Run], it will satisfy the [wizard.ProverAction] interface. It can allow registering a flatten_column for rounds that are larger than 0. For the glue it won't matter but it is an easy way to extend the functionality